lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -1723,12 +1743,24 @@ StructType *UpdateStructTypeForLegacyLayout(StructType *ST,
 
     StructType *NewST =
         StructType::create(ST->getContext(), fieldTypes, legacyName);
-    DxilStructAnnotation *NewSA = TypeSys.AddStructAnnotation(NewST);
-    // Clone annotation.
-    *NewSA = *SA;
-    // Make sure we set the struct type back to the new one, since the
-    // clone would have clobbered it with the old one.
-    NewSA->SetStructType(NewST);
+
+    // Only add annotation if struct is not empty.
+    if (NewST->getNumElements() > 0) {
+      DxilStructAnnotation *NewSA = TypeSys.AddStructAnnotation(NewST);
+
+      // Clone annotation.
```

Comment:
Clone must be expanded manually, since we could be potentially removing fields in the process.